java/ql/lib/semmle/code/java/frameworks/apache/NegativeIOGenerated.qll

```diff
@@ -0,0 +1,769 @@
+        "org.apache.commons.io.monitor;FileAlterationMonitor;getInterval;();generated",
+        "org.apache.commons.io.monitor;FileAlterationMonitor;removeObserver;(FileAlterationObserver);generated",
+        "org.apache.commons.io.monitor;FileAlterationMonitor;start;();generated",
+        "org.apache.commons.io.monitor;FileAlterationMonitor;stop;();generated",
+        "org.apache.commons.io.monitor;FileAlterationMonitor;stop;(long);generated",
+        "org.apache.commons.io.monitor;FileAlterationObserver;checkAndNotify;();generated",
+        "org.apache.commons.io.monitor;FileAlterationObserver;destroy;();generated",
+        "org.apache.commons.io.monitor;FileAlterationObserver;initialize;();generated",
+        "org.apache.commons.io.monitor;FileAlterationObserver;removeListener;(FileAlterationListener);generated",
+        "org.apache.commons.io.monitor;FileEntry;getLastModified;();generated",
+        "org.apache.commons.io.monitor;FileEntry;getLength;();generated",
+        "org.apache.commons.io.monitor;FileEntry;getLevel;();generated",
+        "org.apache.commons.io.monitor;FileEntry;isDirectory;();generated",
+        "org.apache.commons.io.monitor;FileEntry;isExists;();generated",
+        "org.apache.commons.io.monitor;FileEntry;refresh;(File);generated",
+        "org.apache.commons.io.monitor;FileEntry;setDirectory;(boolean);generated",
+        "org.apache.commons.io.monitor;FileEntry;setExists;(boolean);generated",
+        "org.apache.commons.io.monitor;FileEntry;setLastModified;(long);generated",
+        "org.apache.commons.io.monitor;FileEntry;setLength;(long);generated",
+        "org.apache.commons.io.output;AbstractByteArrayOutputStream;AbstractByteArrayOutputStream;();generated",
+        "org.apache.commons.io.output;AbstractByteArrayOutputStream;reset;();generated",
+        "org.apache.commons.io.output;AbstractByteArrayOutputStream;size;();generated",
+        "org.apache.commons.io.output;AbstractByteArrayOutputStream;toByteArray;();generated",
+        "org.apache.commons.io.output;AbstractByteArrayOutputStream;toInputStream;();generated",
+        "org.apache.commons.io.output;AbstractByteArrayOutputStream;write;(InputStream);generated",
+        "org.apache.commons.io.output;AbstractByteArrayOutputStream;writeTo;(OutputStream);generated",
+        "org.apache.commons.io.output;BrokenOutputStream;BrokenOutputStream;();generated",
+        "org.apache.commons.io.output;BrokenOutputStream;BrokenOutputStream;(IOException);generated",
+        "org.apache.commons.io.output;BrokenWriter;BrokenWriter;();generated",
+        "org.apache.commons.io.output;BrokenWriter;BrokenWriter;(IOException);generated",
+        "org.apache.commons.io.output;ByteArrayOutputStream;ByteArrayOutputStream;();generated",
+        "org.apache.commons.io.output;ByteArrayOutputStream;ByteArrayOutputStream;(int);generated",
+        "org.apache.commons.io.output;ByteArrayOutputStream;toBufferedInputStream;(InputStream);generated",
```

Comment:
It may be that I'm misunderstanding negative models, but why should there be no flow here?
> public static [InputStream](http://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html?is-external=true) [toBufferedInputStream](https://commons.apache.org/proper/commons-io/javadocs/api-2.5/src-html/org/apache/commons/io/IOUtils.html#line.518)([InputStream](http://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html?is-external=true) input)
                                         throws [IOException](http://docs.oracle.com/javase/7/docs/api/java/io/IOException.html?is-external=true)
>
> Fetches entire contents of an InputStream and represent same data as result InputStream.

Reply:
Yes, that would seem reasonable. The negative models are generated in the following way:
If the model generator does not discover flow, then we generate a negative model (stating there is no flow). The model generator is far from perfect and this is probably a case, where "positive" flow is missing.
In any case, the negative models are only used for telemetry purposes at the moment, so they don't interfere with dataflow analysis as such.

Reply:
I didn't know that they are only for telemetry. 
Thank you for the explanation.